packages/connect/src/router.ts

```diff
@@ -131,14 +138,42 @@ export function createConnectRouter(
       );
       return this;
     },
-    rpc(service, method, implementation, options) {
-      const { protocols } = whichProtocols(options, base);
-      handlers.push(
-        createUniversalMethodHandler(
-          createMethodImplSpec(service, method, implementation),
-          protocols,
-        ),
-      );
+    rpc(
+      serviceOrMethod: ServiceType | MethodType,
+      methodOrImpl: MethodInfo | MethodImpl<MethodInfo<AnyMessage>>,
+      implementationOrOptions?:
+        | MethodImpl<MethodInfo<AnyMessage>>
+        | Partial<UniversalHandlerOptions>,
+      options?: Partial<UniversalHandlerOptions>,
+    ) {
+      if ("typeName" in serviceOrMethod) {
+        const { protocols } = whichProtocols(options, base);
+
```

Comment:
You will still need to cast, but you can do it at one place:

```ts
      let service: ServiceType;
      let method: MethodInfo;
      let impl: MethodImpl<MethodInfo>;
      let opt: Partial<UniversalHandlerOptions> | undefined;
      if ("typeName" in serviceOrMethod) {
        service = serviceOrMethod;
        method = methodOrImpl as MethodInfo;
       // ...
     } else {
       // ...
     }
     // create handler
```

Then the casts aren't interleaved with creating the handler, and they become a bit easier to reason about.

Reply:
Oh yeah, that is WAY better. I'll update.